mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/impl/MongoDBDatastoreProvider.java

```diff
@@ -168,4 +168,20 @@ private DB extractDatabase(MongoClient mongo, MongoDBConfiguration config) {
 			}
 		}
 	}
+
+	protected MongoDBConfiguration getConfig() {
+		return config;
+	}
+
+	protected MongoClient getMongoClient() {
+		return mongo;
+	}
+
+	protected void setMongoClient(MongoClient mongoClient) {
```

Comment:
To give some background: `DatastoreProvider`s (and other services) are accessed concurrently from several threads at the same time. So their state (fields) must either be immutable or we need to synchronize access to it and make sure changes are visible to other threads (through volatile, synchronized etc.). Of course we strive for avoiding the latter for the sake of performance.

When it comes to immutability, the best is to make state real immutable (by means of setting it in the constructor and use final). But some contracts have a configure() or initialize() method which initializes state after the constructor invocation. Thus we cannot mark those fields as final. Nevertheless no synchronization is required and this is fine as long as these fields are not changed later on again (they are "effectively immutable") as the memory barrier established during building the session factory (which caused the initialization of datastore providers) ensures that these changes from the initialization phase are safely published to other threads.

The same wouldn't apply for a setter call in the course of the program after the initialization phase. Now it may be fine if the setter only actually is called during bootstrap (which I think it is), but that's not implied API-wise. So if we cannot get rid of it, we should at least document the conditions for its usage.